crates/circuit/src/circuit_data.rs

```diff
@@ -129,11 +133,157 @@ pub struct CircuitData {
     /// Mapping between [ShareableClbit] and its locations in
     /// the circuit
     clbit_indices: BitLocator<ShareableClbit, ClassicalRegister>,
+    /// Variables registered in the circuit
+    vars: ObjectRegistry<Var, expr::Var>,
+    /// Stretches registered in the circuit
+    stretches: ObjectRegistry<Stretch, expr::Stretch>,
+    /// Variable identifiers, in order of their addition to the circuit
+    identifier_info: IndexMap<String, CircuitIdentifierInfo>,
+
+    // Var and Stretch indices stored in the circuit
+    vars_input: Vec<Var>,
+    vars_capture: Vec<Var>,
+    vars_declare: Vec<Var>,
+
+    stretches_capture: Vec<Stretch>,
+    stretches_declare: Vec<Stretch>,
```

Comment:
Not for this PR but why do we store the `Var`s and `Stretch`es in `DAGCircuit` within a `HashSet` as opposed to a `Vec`? As far as I'm aware, if anyone needs to find whether a circuit contains a `Var` the `ObjectRegistry` will already give you the ability to quickly check since they implement `Hash`.